File: archive/Cc_prepare_fixations_for_matlab.r
library(tidyverse)

script_path <- rstudioapi::getActiveDocumentContext()$path
dir <- dirname(script_path)
setwd(dir)  #set working directory to script path
subjects = c('02')

blocks = c('11','12','21','22')
for(s in subjects){
  cat(paste(s,"\n"))

  fixations_file = paste(dir,'/fixations/fixations_4analysis.csv',sep='')

  #read fixations file
  fixations = as_tibble(read.csv(fixations_file,sep=',') )

    fixations = fixations %>% filter(!is.na(trial_time) & !is.na(fixation_id))
  names(fixations)[47:62] = c("AA","AB","AC","AD","BA","BB","BC","BD","CA","CB","CC","CD","EA","EB","EC","ED")
  
  
  pattern = 1
  set = 1
  fixations$block = NA
  fixations$fix_at = NA
  fixations$fix_at[!is.na(fixations$fixation_id)]='elsewhere'
    
  fixations$fix_at[fixations$aoi_target == TRUE] = "target"
  fixations$fix_at[fixations$aoi_target == FALSE & (fixations$aoi_comp == TRUE | fixations$aoi_otherTarget == TRUE | fixations$aoi_otherComp == TRUE | fixations$aoi_fillerA == TRUE | fixations$aoi_fillerB == TRUE)] = 'other'
  fixations$fix_at[fixations$aoi_goal == TRUE] = 'goal'

  for(ii in 1:(nrow(fixations)-1)){
    if(pattern > 2){
      pattern = 1
      set = set+1
    }
    else if((fixations$time[ii] < fixations$time[ii+1]) & pattern < 3){
      fixations$block[ii] = paste(set,pattern,sep='')
    }
    else if((fixations$time[ii] > fixations$time[ii+1])  & pattern < 3){
      fixations$block[ii] = paste(set,pattern,sep='')
      pattern = pattern+1
    }
  }
  for(b in blocks){
    cat(paste(b,"..."))
    fixations = fixations[order(fixations$gaze_timestamp),]
    tmp = fixations %>% filter(block == b )
    fix_out =  paste(dir,'/fixations/fixations_times_',b,'_trials.csv',sep='')
    write.csv(file=fix_out,tmp,row.names=FALSE)
    
  }
  
}

File: Cc_prepare_fixations_for_matlab.py
import argparse
import logging
import os
import time

import pandas as pd

from constants import BLOCK_IDS, GAZE_TIMESTAMP_FIELD, WORD_ONSET_FIELD
from load_experiment import (create_experiment_outdir, get_experiment_id,
                             load_config, log_step_duration, parse_subject_ids,
                             subject_dirs_dict)

logger = logging.getLogger(__name__)


ALPHANUMERIC_COLUMN_MAP = {
    '11': 'AA',
    '12': 'AB',
    '13': 'AC',
    '14': 'AD',
    '21': 'BA',
    '22': 'BB',
    '23': 'BC',
    '24': 'BD',
    '31': 'CA',
    '32': 'CB',
    '33': 'CC',
    '34': 'CD',
    '41': 'DA',
    '42': 'DB',
    '43': 'DC',
    '44': 'DD',
}


def main(config: str | dict) -> dict:
    start_time = time.time()
    # Load experiment config
    if isinstance(config, str):
        config = load_config(config)
    experiment_id = get_experiment_id(config)

    # Output paths
    output_dir = create_experiment_outdir(config, experiment_id)
    fixations_dir = config["data"]["input"]["fixations_dir"]
    fixations_outdir = os.path.join(output_dir, fixations_dir)

    # Get selected subject IDs
    _, subject_id_regex = parse_subject_ids(config["experiment"]["subjects"])
    subj_fixation_dirs_dict = subject_dirs_dict(root_dir=fixations_outdir, subject_regex=subject_id_regex)

    # Process per subject

    for subject_id, subj_fixation_dir in subj_fixation_dirs_dict.items():
        if len(subj_fixation_dir) > 1:
            logger.warning(f">1 matching directory found for subject ID '{subject_id}'")
        subj_fixation_dir = subj_fixation_dir[0]
        fixation_file = os.path.join(subj_fixation_dir, "fixations_4analysis.csv")
        subj_fixation_data = pd.read_csv(fixation_file)

        # Filter out NA trial_time or fixation_id
        subj_fixation_data = subj_fixation_data[
            subj_fixation_data["trial_time"].notna() &
            subj_fixation_data["fixation_id"].notna()
        ]

        # Rename columns 11, 12, 13 ... to AA, AB, AC ...
        subj_fixation_data = subj_fixation_data.rename(columns=ALPHANUMERIC_COLUMN_MAP)

        # Annotate block and fix_at columns
        subj_fixation_data["block"] = pd.NA
        subj_fixation_data["fix_at"] = pd.NA
        subj_fixation_data.loc[subj_fixation_data["fixation_id"].notna(), "fix_at"] = "elsewhere"
        subj_fixation_data.loc[
            (subj_fixation_data["aoi_target"] == False) &
            (subj_fixation_data[["aoi_comp", "aoi_otherTarget", "aoi_otherComp", "aoi_fillerA", "aoi_fillerB"]] == True).any(axis=1),
            "fix_at"
        ] = "other"
        subj_fixation_data.loc[subj_fixation_data["aoi_goal"] == True, "fix_at"] = "goal"

        pattern_id, set_id = 1, 1
        for idx in range(len(subj_fixation_data) - 1):
            current_time = subj_fixation_data.iloc[idx][WORD_ONSET_FIELD]
            next_time = subj_fixation_data.iloc[idx + 1][WORD_ONSET_FIELD]

            if pattern_id > 2:
                pattern_id = 1
                set_id += 1

            if current_time < next_time and pattern_id < 3:
                subj_fixation_data.iloc[idx, subj_fixation_data.columns.get_loc("block")] = f"{set_id}{pattern_id}"
            elif current_time > next_time and pattern_id < 3:
                subj_fixation_data.iloc[idx, subj_fixation_data.columns.get_loc("block")] = f"{set_id}{pattern_id}"
                pattern_id += 1

        # Sort by gaze_timestamp field
        subj_fixation_data = subj_fixation_data.sort_values(by=GAZE_TIMESTAMP_FIELD)

        # Iterate over blocks and write filtered block data to CSV
        for block in BLOCK_IDS:
            block_data = subj_fixation_data[subj_fixation_data["block"] == str(block)]
            fixation_outfile = os.path.join(subj_fixation_dir, f"fixations_times_{block}_trials.csv")
            block_data.to_csv(fixation_outfile, index=False)
        logger.info(f"Wrote subject <{subject_id}> block fixation files to {subj_fixation_dir}")

    # Log duration of this step in run config
    log_step_duration(config, start_time, step_id="Cc_prepare_fixations_for_matlab")

    return config


if __name__ == "__main__":
    parser = argparse.ArgumentParser("Prepare fixations for inclusion in EEG.event structure in MATLAB")
    parser.add_argument('config', help='Path to config.yml file')
    args = parser.parse_args()
    main(args.config)
